File: apps/web/lib/services/asset-service.ts
const API_BASE_URL = process.env.NEXT_PUBLIC_ASSET_SERVICE_URL || 'http://localhost:3003';

export interface Asset {
  id: string;
  assetCode: string;
  name: string;
  description?: string;
  model?: string;
  manufacturer?: string;
  serialNumber?: string;
  location: string;
  isActive: boolean;
  ownerId: string;
  owner: { firstName: string; lastName: string };
  administratorId: string;
  administrator: { firstName: string; lastName: string };
  installDate?: string;
  createdAt: string;
  updatedAt: string;
}

export interface PaginatedAssets {
  assets: Asset[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export interface AssetStats {
  total: number;
  active: number;
  inactive: number;
  locations: number;
  byLocation: Record<string, number>;
  byManufacturer: Record<string, number>;
}

interface AssetKPIFilters {
  timeRange?: string;
  location?: string;
  assetType?: string;
  startDate?: string;
  endDate?: string;
  limit?: number;
}

interface AssetFilters {
  location?: string;
  isActive?: boolean;
  search?: string;
  page?: number;
  limit?: number;
}

class AssetService {
  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const token = localStorage.getItem('auth_token');
    
    const response = await fetch(`${API_BASE_URL}${endpoint}`, {
      headers: {
        'Content-Type': 'application/json',
        ...(token && { Authorization: `Bearer ${token}` }),
        ...options.headers,
      },
      ...options,
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({ message: 'Network error' }));
      throw new Error(errorData.message || `HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    return data.data || data;
  }

  private buildQueryParams(filters: AssetKPIFilters | AssetFilters): string {
    const params = new URLSearchParams();
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        params.append(key, value.toString());
      }
    });
    return params.toString();
  }

  // Basic CRUD operations
  async getAllAssets(filters: AssetFilters = {}): Promise<PaginatedAssets> {
    const queryParams = this.buildQueryParams(filters);
    const token = localStorage.getItem('auth_token');
    
    const response = await fetch(`${API_BASE_URL}/api/assets?${queryParams}`, {
      headers: {
        'Content-Type': 'application/json',
        ...(token && { Authorization: `Bearer ${token}` }),
      },
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({ message: 'Network error' }));
      throw new Error(errorData.message || `HTTP error! status: ${response.status}`);
    }

    const result = await response.json();
    
    // Handle the specific backend response format for assets list
    if (result.success && result.data && result.pagination) {
      return {
        assets: result.data,
        total: result.pagination.total,
        page: result.pagination.page,
        limit: result.pagination.limit,
        totalPages: result.pagination.totalPages,
      };
    }
    
    // Fallback to direct data if format is different
    return result.data || result;
  }

  async getAssetById(id: string): Promise<Asset> {
    return this.request<Asset>(`/api/assets/${id}`);
  }

  async createAsset(assetData: Partial<Asset>): Promise<Asset> {
    return this.request<Asset>('/api/assets', {
      method: 'POST',
      body: JSON.stringify(assetData),
    });
  }

  async updateAsset(id: string, assetData: Partial<Asset>): Promise<Asset> {
    return this.request<Asset>(`/api/assets/${id}`, {
      method: 'PUT',
      body: JSON.stringify(assetData),
    });
  }

  async deleteAsset(id: string): Promise<void> {
    return this.request<void>(`/api/assets/${id}`, {
      method: 'DELETE',
    });
  }

  // Search and stats
  async searchAssets(query: string): Promise<{ assets: Asset[]; total: number }> {
    return this.request<{ assets: Asset[]; total: number }>(`/api/assets/search?q=${encodeURIComponent(query)}`);
  }

  // Manual asset code input methods
  async searchAssetsByCode(partialCode: string, filters: { location?: string; isActive?: boolean; limit?: number } = {}): Promise<Asset[]> {
    const params = new URLSearchParams();
    params.append('code', partialCode);
    
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        params.append(key, value.toString());
      }
    });

    return this.request<Asset[]>(`/api/assets/search-by-code?${params.toString()}`);
  }

  async validateAssetCode(assetCode: string): Promise<{ exists: boolean; asset?: Asset }> {
    const params = new URLSearchParams();
    params.append('code', assetCode);
    
    return this.request<{ exists: boolean; asset?: Asset }>(`/api/assets/validate?${params.toString()}`);
  }

  async getAssetSuggestions(input: string, filters: { location?: string; isActive?: boolean; limit?: number } = {}): Promise<Asset[]> {
    const params = new URLSearchParams();
    params.append('input', input);
    
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        params.append(key, value.toString());
      }
    });

    return this.request<Asset[]>(`/api/assets/suggest?${params.toString()}`);
  }

  async getAssetStats(): Promise<AssetStats> {
    const token = localStorage.getItem('auth_token');
    
    const response = await fetch(`${API_BASE_URL}/api/assets/stats`, {
      headers: {
        'Content-Type': 'application/json',
        ...(token && { Authorization: `Bearer ${token}` }),
      },
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({ message: 'Network error' }));
      throw new Error(errorData.message || `HTTP error! status: ${response.status}`);
    }

    const result = await response.json();
    return result.data || result;
  }

  async getLocations(): Promise<{ locations: string[] }> {
    const token = localStorage.getItem('auth_token');
    
    const response = await fetch(`${API_BASE_URL}/api/assets/locations`, {
      headers: {
        'Content-Type': 'application/json',
        ...(token && { Authorization: `Bearer ${token}` }),
      },
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({ message: 'Network error' }));
      throw new Error(errorData.message || `HTTP error! status: ${response.status}`);
    }

    const result = await response.json();
    return result.data || result;
  }

  // KPI methods (keeping existing ones)
  async getDowntimeRanking(filters: AssetKPIFilters = {}): Promise<any[]> {
    const queryParams = this.buildQueryParams(filters);
    return this.request<any[]>(`/api/assets/kpi/downtime-ranking?${queryParams}`);
  }

  async getFaultFrequencyRanking(filters: AssetKPIFilters = {}): Promise<any[]> {
    const queryParams = this.buildQueryParams(filters);
    return this.request<any[]>(`/api/assets/kpi/fault-frequency?${queryParams}`);
  }

  async getMaintenanceCostAnalysis(filters: AssetKPIFilters = {}): Promise<any[]> {
    const queryParams = this.buildQueryParams(filters);
    return this.request<any[]>(`/api/assets/kpi/maintenance-cost?${queryParams}`);
  }

  async getHealthOverview(filters: AssetKPIFilters = {}): Promise<any> {
    const queryParams = this.buildQueryParams(filters);
    return this.request<any>(`/api/assets/kpi/health-overview?${queryParams}`);
  }

  async getPerformanceRanking(filters: AssetKPIFilters = {}): Promise<any[]> {
    const queryParams = this.buildQueryParams(filters);
    return this.request<any[]>(`/api/assets/kpi/performance-ranking?${queryParams}`);
  }

  async getCriticalAssets(filters: AssetKPIFilters = {}): Promise<any[]> {
    const queryParams = this.buildQueryParams(filters);
    return this.request<any[]>(`/api/assets/kpi/critical-assets?${queryParams}`);
  }
}

export const assetService = new AssetService();

File: apps/mobile/lib/shared/services/asset_service.dart
import '../models/asset.dart';
import 'api_client.dart';

class AssetValidationResult {
  final bool exists;
  final Asset? asset;

  const AssetValidationResult({
    required this.exists,
    this.asset,
  });
}

class AssetService {
  late final ApiClient _apiClient;
  late final ApiClient _assetServiceClient;
  
  static AssetService? _instance;
  
  AssetService._internal();
  
  static Future<AssetService> getInstance() async {
    if (_instance == null) {
      _instance = AssetService._internal();
      // 使用用户服务客户端，因为asset相关API在user service中
      _instance!._apiClient = await ApiClient.getUserServiceClient();
      // 为手工输入功能使用专门的资产服务客户端
      _instance!._assetServiceClient = await ApiClient.getAssetServiceClient();
    }
    return _instance!;
  }
  
  Future<Asset> getAssetByCode(String assetCode) async {
    
    // 检查token状态
    final token = await _apiClient.getToken();
    if (token != null) {
    }
    
    try {
      final response = await _apiClient.get<Map<String, dynamic>>(
        '/api/assets/code/$assetCode',
      );
      
      
      if (response.data == null) {
        throw const ApiException(message: '设备信息响应数据为空');
      }
      
      // 检查响应格式，如果是包装在success字段中的数据
      final Map<String, dynamic> responseData = response.data!;
      Map<String, dynamic> assetData;
      
      if (responseData.containsKey('success') && responseData['success'] == true) {
        // 标准API响应格式：{"success": true, "data": {...}}
        assetData = responseData['data'] as Map<String, dynamic>;
      } else {
        // 直接响应格式
        assetData = responseData;
      }
      
      return Asset.fromJson(assetData);
      
    } catch (e) {
      rethrow;
    }
  }

  // Manual asset code input methods
  Future<List<Asset>> searchAssetsByCode(
    String partialCode, {
    String? location,
    bool? isActive,
    int limit = 10,
  }) async {
    
    final queryParameters = <String, dynamic>{
      'code': partialCode,
      'limit': limit,
    };
    
    if (location != null && location.isNotEmpty) {
      queryParameters['location'] = location;
    }
    
    if (isActive != null) {
      queryParameters['status'] = isActive ? 'ACTIVE' : 'INACTIVE';
    }
    
    try {
      final response = await _assetServiceClient.get<Map<String, dynamic>>(
        '/api/assets/search-by-code',
        queryParameters: queryParameters,
      );
      
      
      if (response.data == null || response.data!['success'] != true) {
        return [];
      }
      
      final List<dynamic> assetsData = response.data!['data'] as List<dynamic>;
      return assetsData
          .map((json) => Asset.fromJson(json as Map<String, dynamic>))
          .toList();
      
    } catch (e) {
      return [];
    }
  }

  Future<AssetValidationResult> validateAssetCode(String assetCode) async {
    
    try {
      final response = await _assetServiceClient.get<Map<String, dynamic>>(
        '/api/assets/validate',
        queryParameters: {'code': assetCode},
      );
      
      
      if (response.data == null) {
        return AssetValidationResult(exists: false);
      }
      
      final bool exists = response.data!['exists'] == true;
      Asset? asset;
      
      if (exists && response.data!['asset'] != null) {
        asset = Asset.fromJson(response.data!['asset'] as Map<String, dynamic>);
      }
      
      return AssetValidationResult(exists: exists, asset: asset);
      
    } catch (e) {
      return AssetValidationResult(exists: false);
    }
  }

  Future<List<Asset>> getAssetSuggestions(
    String input, {
    String? location,
    bool? isActive,
    int limit = 10,
  }) async {
    
    final queryParameters = <String, dynamic>{
      'input': input,
      'limit': limit,
    };
    
    if (location != null && location.isNotEmpty) {
      queryParameters['location'] = location;
    }
    
    if (isActive != null) {
      queryParameters['status'] = isActive ? 'ACTIVE' : 'INACTIVE';
    }
    
    try {
      final response = await _assetServiceClient.get<Map<String, dynamic>>(
        '/api/assets/suggest',
        queryParameters: queryParameters,
      );
      
      
      if (response.data == null || response.data!['success'] != true) {
        return [];
      }
      
      final List<dynamic> assetsData = response.data!['data'] as List<dynamic>;
      return assetsData
          .map((json) => Asset.fromJson(json as Map<String, dynamic>))
          .toList();
      
    } catch (e) {
      return [];
    }
  }
  
  Future<Asset> getAssetById(String id) async {
    final response = await _apiClient.get<Map<String, dynamic>>(
      '/api/assets/$id',
    );
    
    if (response.data == null) {
      throw const ApiException(message: '设备信息响应数据为空');
    }
    
    return Asset.fromJson(response.data!);
  }
  
  Future<List<Asset>> getAssets({
    int page = 1,
    int limit = 20,
    String? search,
    String? location,
  }) async {
    final queryParameters = <String, dynamic>{
      'page': page,
      'limit': limit,
    };
    
    if (search != null && search.isNotEmpty) {
      queryParameters['search'] = search;
    }
    
    if (location != null && location.isNotEmpty) {
      queryParameters['location'] = location;
    }
    
    final response = await _apiClient.get<Map<String, dynamic>>(
      '/api/assets',
      queryParameters: queryParameters,
    );
    
    if (response.data == null) {
      throw const ApiException(message: '设备列表响应数据为空');
    }
    
    final List<dynamic> assetsJson = response.data!['assets'] as List<dynamic>;
    return assetsJson
        .map((json) => Asset.fromJson(json as Map<String, dynamic>))
        .toList();
  }
  
  Future<List<String>> getAssetCategories() async {
    // This could be retrieved from a backend endpoint
    // For now, return predefined categories
    return [
      '电气设备',
      '机械设备',
      '自动化设备',
      '安全设备',
      '环境设备',
      '通信设备',
      '计算机设备',
      '办公设备',
      '其他',
    ];
  }
  
  Future<List<String>> getFailureReasons() async {
    // This could be retrieved from a backend endpoint
    // For now, return predefined reasons
    return [
      '设备故障',
      '性能下降',
      '异常噪音',
      '温度异常',
      '泄漏',
      '磨损',
      '电气故障',
      '软件故障',
      '定期维护',
      '预防性维护',
      '其他',
    ];
  }
  
  Future<List<String>> getCommonLocations() async {
    // This could be retrieved from a backend endpoint
    // For now, return predefined locations
    return [
      '生产车间A',
      '生产车间B',
      '生产车间C',
      '仓库',
      '办公区',
      '设备间',
      '电气房',
      '机房',
      '实验室',
      '其他',
    ];
  }
}